python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,154 @@
+
+    Args:
+        location(str): A URI or a path to a local file
+
+    Attributes:
+        location(str): The URI or path to a local file for a PyArrowFile instance
+        exists(bool): Whether the file exists or not
+        filesystem(pyarrow.fs.FileSystem): An implementation of the FileSystem base class inferred from the location
+
+    Examples:
+        >>> from iceberg.io.pyarrow import PyArrowFile
+        >>> input_file = PyArrowFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()  # Read the contents of the PyArrowFile instance
+        >>> output_file = PyArrowFile("s3://baz/qux.txt")
+        >>> output_file.create().write(b'foobytes')  # Write bytes to the PyArrowFile instance
+    """
+
+    def __init__(self, location: str):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+        if not parsed_location.scheme:  # If no scheme, assume the path is to a local file
+            self._filesystem, self._path = FileSystem.from_uri(os.path.abspath(location))
+        else:
+            self._filesystem, self._path = FileSystem.from_uri(location)  # Infer the proper filesystem
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file = self._filesystem.open_input_file(self._path)
+        return file.size()
+
+    @property
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
+        file_info = self._filesystem.get_file_info(self._path)
+        return False if file_info.type == FileType.NotFound else True
+
+    def open(self) -> InputStream:
+        """Opens the location using a PyArrow FileSystem inferred from the location
+
+        Returns:
+            pyarrow.lib.NativeFile: A NativeFile instance for the file located at `self.location`
+        """
+        input_file = self._filesystem.open_input_file(self._path)
+        if not isinstance(input_file, InputStream):
+            raise TypeError(
+                f"Object of type {type(input_file)} returned from PyArrowFile.open does not match the InputStream protocol."
+            )
+        return input_file
+
+    def create(self, overwrite: bool = False) -> OutputStream:
+        """Creates a writable pyarrow.lib.NativeFile for this PyArrowFile's location
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            pyarrow.lib.NativeFile: A NativeFile instance for the file located at self.location
+
+        Raises:
+            FileExistsError: If the file already exists at `self.location` and `overwrite` is False
+        """
+        if not overwrite and self.exists:
+            raise FileExistsError(
+                f"A file already exists at this location. If you would like to overwrite it, set `overwrite=True`: {self.location}"
+            )
+        output_file = self._filesystem.open_output_stream(self._path)
```

Comment:
If opening fails because the file was created concurrently, does it arrow raise `FileExistsError`?

Reply:
Did you mean if opening the file succeeds? For `open_output_stream`, the docstring says `If the target already exists, existing data is truncated.`. That would be a problem here where if `self.exists()` on the line above returns `False`, then a concurrent process creates and writes the files, then the `open_output_stream` here would successfully open the output stream but would truncate the contents. So it would effectively overwrite it even though `overwrite=False`.

@emkornfield, is there a way with pyarrow to open an output stream in a way that fails if the file already exists?

Reply:
I see. So this is always an overwrite? That seems like a problem to me, but we may not be able to do anything about it besides have the existence check above. At a minimum, we should note that this can overwrite concurrently created files.